[FEAT] 마이 프로필 및 회원 탈퇴

stopstone opened the pull request:
## 작업 내용
- MY 클릭 시 마이페이지 이동
- 서비스 이용에 필요한 약관 및 설명을 webView로 구성하였습니다.
- DataStore 초기화로 로그아웃, 회원 탈퇴는 api요청으로 동작합니다.
- 탈퇴사유에서 텍스트필드는 NearOutlinedTextField 를 사용합니다.
- 기타인 경우만 사유를 입력할 수 있습니다.

## 확인 방법
- feature/my-profile에서 MyProfile~, Withdraw~ 확인 부탁드립니다!
- NearOutlinedTextField 를 새로 구성하였습니다. ui/components에 위치하였습니다.
- 웹뷰 스크린 WebViewFrame은 ui/components로 구성하였습니다.

## 참고 사항
- 텍스트필드를 포커스별 border를 조정, 에러 메시지 설정, 글자수 설정 등으로 변경사항이 많을 것 같다
새롭게 구성하였습니다!
- 웹뷰 구성으로 local.properties에 약관 링크를 넣어야합니다! 디스코드 참고 부탁드려요!
- 회원 탈퇴시 다이얼로그로 확인하지 않고 바로 탈퇴를 하고 있습니다! 다이얼로그 디자인이 나오면 수정하겠습니다

## 관련 이슈
- close #9

stopstone committed:
feat: 온보딩 버튼 컴포넌트 추가

- ui/components의 NearBasicButton을 기반하여 만들었습니다.

- 현재 페이지가 마지막 페이지인지에 따라 버튼 텍스트를 "로그인/회원가입" 또는 "다음"으로 표시합니다.

- 버튼 클릭 시 `onNextClick` 콜백을 호출합니다.

stopstone committed:
feat: 온보딩 버튼 컴포넌트 추가

- ui/components의 NearBasicButton을 기반하여 만들었습니다.

- 현재 페이지가 마지막 페이지인지에 따라 버튼 텍스트를 "로그인/회원가입" 또는 "다음"으로 표시합니다.

- 버튼 클릭 시 `onNextClick` 콜백을 호출합니다.

stopstone committed:
feat: 온보딩 페이지 인디케이터 UI 구현

- 현재 페이지의 컬러를 BLUE01_5AA2E9로 설정
- 이외 컬리는 GRAY03_EBEBEB로 설정합니다.

stopstone committed:
feat: 온보딩 페이지 인디케이터 UI 구현

- 현재 페이지의 컬러를 BLUE01_5AA2E9로 설정
- 이외 컬리는 GRAY03_EBEBEB로 설정합니다.

stopstone committed:
feat: 온보딩 페이지 모델 구현

- 각각의 페이지의 구성요소를 담당하는 data class입니다.
- 추후 이미지를 추가하여 온보딩 페이지에서 이미지를 그리는 역할을 합니다.

stopstone committed:
feat: 온보딩 페이지 모델 구현

- 각각의 페이지의 구성요소를 담당하는 data class입니다.
- 추후 이미지를 추가하여 온보딩 페이지에서 이미지를 그리는 역할을 합니다.

stopstone committed:
feat: 온보딩 화면 구현

- 각 컴포넌트를 조합하여 온보딩 화면을 구현하였습니다.

- createAnnotatedText() 함수를 통해 \n 이후 텍스트를 강조합니다.

- stringResource를 통해 페이지 별 타이틀을 string.xml에서 가져옵니다.

stopstone committed:
feat: 온보딩 화면 구현

- 각 컴포넌트를 조합하여 온보딩 화면을 구현하였습니다.

- createAnnotatedText() 함수를 통해 \n 이후 텍스트를 강조합니다.

- stringResource를 통해 페이지 별 타이틀을 string.xml에서 가져옵니다.

stopstone committed:
feat: 온보딩 네비게이션 설정

stopstone committed:
feat: 온보딩 네비게이션 설정

stopstone committed:
feat: 온보딩 페이지 이미지 추가

stopstone committed:
feat: 온보딩 페이지 이미지 추가

stopstone committed:
feat: 온보딩 화면 이미지 추가

stopstone committed:
feat: 온보딩 화면 이미지 추가

stopstone committed:
feat: 온보딩 UI 이미지 영역 수정

stopstone committed:
feat: 온보딩 UI 이미지 영역 수정

stopstone committed:
feat: 온보딩 페이지 백그라운드 구성

stopstone committed:
feat: 온보딩 페이지 백그라운드 구성

stopstone committed:
feat: 온보딩 배경 추가

stopstone committed:
feat: 온보딩 배경 추가

stopstone committed:
refactor: 불필요한 패딩 제거

stopstone committed:
refactor: 불필요한 패딩 제거

stopstone committed:
feat: 온보딩 백그라운드 색상 변경

stopstone committed:
feat: 온보딩 백그라운드 색상 변경

stopstone committed:
refactor: OnboardingScreen UI 레이아웃 조정

OnboardingScreen의 UI 레이아웃을 수정하여 상태바와 네비게이션 바 영역을 고려하도록 변경했습니다.
`systemBarsPadding()` 대신 `WindowInsets`를 사용하여 직접 패딩을 계산하고 적용했습니다.

stopstone committed:
refactor: OnboardingScreen UI 레이아웃 조정

OnboardingScreen의 UI 레이아웃을 수정하여 상태바와 네비게이션 바 영역을 고려하도록 변경했습니다.
`systemBarsPadding()` 대신 `WindowInsets`를 사용하여 직접 패딩을 계산하고 적용했습니다.

stopstone committed:
refactor: AnnotatedString.Builder 확장 함수 추가

stopstone committed:
refactor: AnnotatedString.Builder 확장 함수 추가

stopstone committed:
feat: DataStore 분리 및 Qualifier 적용

stopstone committed:
feat: DataStore 분리 및 Qualifier 적용

stopstone committed:
feat: 온보딩 DataStore 추가

온보딩 완료 상태를 저장하고 조회하는 `OnboardingPreferences` 클래스를 추가했습니다.
이를 통해 앱 최초 실행 시 온보딩 화면 표시 여부를 결정할 수 있습니다.

stopstone committed:
feat: 온보딩 DataStore 추가

온보딩 완료 상태를 저장하고 조회하는 `OnboardingPreferences` 클래스를 추가했습니다.
이를 통해 앱 최초 실행 시 온보딩 화면 표시 여부를 결정할 수 있습니다.

stopstone committed:
feat: 온보딩 완료 상태 저장 로직 추가

stopstone committed:
feat: 온보딩 완료 상태 저장 로직 추가

stopstone committed:
feat: 온보딩 버튼 로딩 상태 추가

- 온보딩 버튼에 로딩 상태를 추가했습니다.
- 로딩 중일 때는 버튼이 비활성화됩니다.

stopstone committed:
feat: 온보딩 버튼 로딩 상태 추가

- 온보딩 버튼에 로딩 상태를 추가했습니다.
- 로딩 중일 때는 버튼이 비활성화됩니다.

stopstone committed:
feat: 온보딩 화면 UI 상태 및 사이드 이펙트 처리

- OnboardingViewModel에 UiState 및 Effect 추가
- OnboardingScreen에서 UiState 관찰 및 Effect 처리 로직 구현
- 온보딩 완료 시 로딩 상태 표시 및 에러 처리 추가

stopstone committed:
feat: 온보딩 화면 UI 상태 및 사이드 이펙트 처리

- OnboardingViewModel에 UiState 및 Effect 추가
- OnboardingScreen에서 UiState 관찰 및 Effect 처리 로직 구현
- 온보딩 완료 시 로딩 상태 표시 및 에러 처리 추가

stopstone committed:
feat: 온보딩 버튼 텍스트 stirng.xml 리소스 추가

stopstone committed:
feat: 온보딩 버튼 텍스트 stirng.xml 리소스 추가

stopstone committed:
feat: 온보딩 레포지토리 추가 및 ViewModel 연동

- 온보딩 완료 여부 저장 및 조회 기능을 위한 `OnBoardingRepository` 인터페이스와 `OnBoardingRepositoryImpl` 구현체를 추가했습니다.
- `OnboardingViewModel`이 `OnboardingPreferences` 대신 `OnBoardingRepository`를 사용하도록 수정했습니다.
- `RepositoryModule`에 `OnBoardingRepository` 관련 의존성 주입 설정을 추가했습니다.

stopstone committed:
feat: 온보딩 레포지토리 추가 및 ViewModel 연동

- 온보딩 완료 여부 저장 및 조회 기능을 위한 `OnBoardingRepository` 인터페이스와 `OnBoardingRepositoryImpl` 구현체를 추가했습니다.
- `OnboardingViewModel`이 `OnboardingPreferences` 대신 `OnBoardingRepository`를 사용하도록 수정했습니다.
- `RepositoryModule`에 `OnBoardingRepository` 관련 의존성 주입 설정을 추가했습니다.

stopstone committed:
refactor: 온보딩 페이지 데이터 - remember로 성능 최적화

stopstone committed:
refactor: 온보딩 페이지 데이터 - remember로 성능 최적화

stopstone committed:
refactor: 온보딩 Preferences 키 상수화

stopstone committed:
refactor: 온보딩 Preferences 키 상수화

stopstone committed:
refactor: 불필요한 코드 제거

stopstone committed:
refactor: 불필요한 코드 제거

stopstone committed:
refactor: 온보딩 페이지 데이터 모델 변경

- `title` 필드를 `String`에서 `@StringRes` 어노테이션을 사용하는 `Int` 타입의 `titleResId`로 변경했습니다.
- 이에 따라 `OnboardingScreen`에서 `title`을 참조하는 부분을 `stringResource`를 사용하도록 수정했습니다.

stopstone committed:
refactor: 온보딩 페이지 데이터 모델 변경

- `title` 필드를 `String`에서 `@StringRes` 어노테이션을 사용하는 `Int` 타입의 `titleResId`로 변경했습니다.
- 이에 따라 `OnboardingScreen`에서 `title`을 참조하는 부분을 `stringResource`를 사용하도록 수정했습니다.

stopstone committed:
refactor: 디자인 가이드에 맞게 텍스트 수정

stopstone committed:
refactor: 디자인 가이드에 맞게 텍스트 수정

stopstone committed:
feat: Color.kt 온보딩 배경 장식 색상 추가

stopstone committed:
feat: Color.kt 온보딩 배경 장식 색상 추가

stopstone committed:
refactor: 불필요한 import문 제거

stopstone committed:
refactor: 불필요한 import문 제거

stopstone committed:
feat: 마이페이지 구성에 필요한 리소스 추가

stopstone committed:
feat: LoginType enum 추가

- ProviderType에 따라 텍스트와 아이콘을 가진다.

stopstone committed:
feat: 마이프로필 컴포넌트 추가

- NearLogoutButton: 로그아웃 버튼
- NearServiceInfoRow: 서비스 정보 행
- NearSocialLoginBadge: 소셜 로그인 뱃지
- NearSwitch: 토글 스위치

stopstone committed:
feat: 내 정보 화면 UI 및 ViewModel 구현

stopstone committed:
refactor: LoginType model 패키지로 이동

stopstone committed:
feat: 내 프로필 화면 navigation 구현

- 홈에서 MY 클릭 시 마이페이지 이동

stopstone committed:
feat: Glide Compose 라이브러리 추가

stopstone committed:
feat: 회원 정보 API 연동

stopstone committed:
feat: MyProfile 에러 핸들링 추가

stopstone committed:
feat: 프로필 이미지 로딩 구현

stopstone committed:
feat: 멤버 정보 API 연동

- MemberRepository 인터페이스 및 구현체 추가
- getMyInfo API 호출 로직 구현
- RepositoryModule에 MemberRepository 바인딩

stopstone committed:
feat: 내 정보 화면 및 API 연동

- 내 정보 화면 UI 구현
- 내 정보 API 연동 및 ViewModel, UiState, UiEvent 정의
- 사용자 닉네임, 로그인 타입 표시
- 알림 설정, 로그아웃, 회원탈퇴 기능 UI 구성

stopstone committed:
feat: 마이프로필 뒤로가기 기능 구현

- MyProfileViewModel에 UI 이벤트 채널 및 관련 로직 추가
- MyProfileScreen에서 뒤로가기 버튼 클릭 시 내비게이션 이벤트 처리
- MyProfileScreen의 NearTopAppbar에 뒤로가기 버튼 핸들러 연결

stopstone committed:
refactor: 마이프로필 정보에 매퍼 적용 및 UI 연결

- MemberInfo를 MyProfileInfo로 변환하는 매퍼 추가
- MyProfileViewModel에서 MemberInfo 대신 MyProfileInfo 사용하도록 수정
- MyProfileScreen에서 MyProfileInfo를 사용하여 UI 업데이트

stopstone committed:
feat: 마이프로필 로그아웃 기능 구현

- MyProfileViewModel에 authRepository를 주입하여 로그아웃 로직을 추가했습니다.
- 로그아웃 성공 시 로그인 화면으로 이동하도록 NavigationMyProfile, MyProfileScreen, NearNavHost를 수정했습니다.
- MyProfileUiEvent에 Logout 이벤트를 추가했습니다.

stopstone committed:
refactor: 로딩 화면 범위 수정

stopstone committed:
refactor: MyProfileScreen 부분별 코드 분리

- MyProfileInfoSection(): 프로플 정보를 나타내는 부분
- MyProfileGeneralSection(): 연결계정, 알림설정 및 계쩡 관련 부분
- MyProfileInfoSection(): Near앱 서비스 고지 관련 부분

stopstone committed:
feat: 회원탈퇴 화면 이동 로직 구현

stopstone committed:
feat: 회원탈퇴 API 연동

stopstone committed:
feat: 회원탈퇴 사유 모델 추가

- WithdrawReason enum 클래스 추가
- WithdrawReasonItem 데이터 클래스 추가

stopstone committed:
feat: 회원탈퇴 화면 TopAppBar 구현

- 백 버튼 없이 취소 버튼만 있는 `WithdrawTopAppBar` Composable을 추가했습니다.
- `title`과 `onCancelClick`을 파라미터로 받아 화면 제목과 취소 버튼 클릭 시 동작을 설정할 수 있습니다.

stopstone committed:
feat: 회원탈퇴 화면 및 기능 구현

- 회원탈퇴 화면 UI를 구성했습니다.
- 회원탈퇴 이유 선택 및 기타 이유 입력 기능을 추가했습니다.
- 회원탈퇴 API 연동 및 로그아웃 후 로그인 화면 이동 로직을 구현했습니다.
- 회원탈퇴 화면 상단바에 사용될 취소 아이콘(ic_32_cancel)을 추가했습니다.

stopstone committed:
feat: NearOutlinedTextField 컴포넌트 추가

- 커스텀 `OutlinedTextField` 구현
- 글자 수 표시 기능 추가
- 다양한 상태에 대한 Preview 추가
- TODO 키보드와 텍스트필드 사이 패딩 필요

stopstone committed:
refactor: NearOutlinedTextField 키보드 상호작용 개선

- 키보드 위에 텍스트필드를 배치
- 글자가 입력될 때 텍스트필드 사이즈를 반영하여 키보드 위에 배치
- 패딩값 동적으로 계산하여 반영

stopstone committed:
feat: 텍스트필드 에러 상태 및 관련 기능 추가

- 텍스트필드에 `isError` 상태를 추가하고, 에러 발생 시 에러 메시지를 표시하도록 구현했습니다.
- 에러가 발생하면 해당 뷰로 자동 스크롤되는 기능을 추가했습니다.
- `focusRequester`를 외부에서 주입받아 포커스를 제어할 수 있도록 수정했습니다.
- 한 번이라도 입력이 시작되면 글자 수가 항상 표시되도록 로직을 변경했습니다.

stopstone committed:
feat: 회원탈퇴 화면 UI 및 로직 개선

- `WithdrawUiState`에 탈퇴하기 버튼 활성화 및 기타 사유 입력 로직을 관리하는 상태 프로퍼티를 추가했습니다.
- '기타' 사유 선택 시, 키보드가 먼저 올라오도록 포커스를 요청하고 텍스트 필드를 초기화하는 로직을 추가했습니다.
- 기타 사유를 입력할 수 있는 `NearOutlinedTextField`를 추가하고, 선택 여부에 따라 활성화 상태를 제어합니다.

stopstone committed:
style: 회원탈퇴 화면 앱바 패딩 수정

stopstone committed:
refactor: 의견 입력 에러 메시지 수정

stopstone committed:
feat: 회원탈퇴 화면 뒤로가기 추가

stopstone committed:
feat: 회원탈퇴 화면에 닉네임 전달 로직 추가

- saveStateHandle로 닉네임 값을 넘겨받아 닉네임을 UI에 보여주기

stopstone committed:
chore: 탈퇴 화면 하단 패딩 수정

stopstone committed:
chore: NearServiceInfoRow 클릭 범위 수정 및 리플효과 제거

stopstone committed:
refactor: WithdrawTopAppBar → NearCancelTopAppBar로 이름 변경 및 이동

- 공통 컴포넌트로 활용하기 위해 ui/componenets로 이동

stopstone committed:
chore: 미사용 import 제거

stopstone committed:
refactor: Modifier 파라미터 적용

stopstone committed:
chore: build.gradle.kts 약관 URL 정보 추가

stopstone committed:
feat: 약관 및 정책 웹뷰 화면 연동

- 약관 및 정책의 제목과 URL을 관리하는 `TermsType` enum을 추가했습니다.
- 마이프로필 화면에서 약관 및 정책 항목을 클릭하면 해당 URL을 웹뷰로 여는 기능을 추가했습니다.
- 웹뷰를 표시하기 위한 `WebView` composable과 관련 네비게이션 로직을 추가했습니다.

stopstone committed:
refactor: 약관 및 정책 목록 표시 로직 개선

- TermsType enum class를 활용하여 NearServiceInfoRow를 반복하여 호출하였습니다.

stopstone committed:
refactor: 약관 페이지 title 형식 수정

- title 뒤에 "상세" 문자열 추가

stopstone committed:
feat: MY 페이지, 탈퇴하기 화면 string.xml 리소스 적용

- MY 페이지, 탈퇴하기 화면에서 사용할 스트링 리소스를 string.xml로 관리합니다.
- 하드코딩된 문자열을 `stringResource`를 사용하여 스트링 리소스로 대체했습니다.
- `LoginType`, `TermsType`, `WithdrawReason` enum 클래스가 스트링 리소스 ID를 갖도록 수정했습니다.

stopstone committed:
remove: 회원 탈퇴 사유 모델 제거

stopstone committed:
refactor: LoginType의 프로퍼티명 수정

stopstone committed:
chore: 내 정보 화면 배경색 추가

stopstone committed:
feat: 앱 시작 시 온보딩 및 로그인 상태에 따른 화면 분기 처리

- MainViewModel에 onBoardingRepository를 주입하여 온보딩 완료 여부를 확인합니다.
- MainActivity, NearApp, NearNavHost에서 로그인 상태(isLoggedIn) 대신 시작 화면(startDestination)을 사용하도록 수정했습니다.
- MainUiState에 isOnboardingCompleted와 startDestination 필드를 추가했습니다.
- 앱 시작 시 온보딩 완료 여부와 로그인 상태를 확인하여 적절한 화면(온보딩, 홈, 로그인)으로 이동하도록 MainViewModel의 checkAppStatus 로직을 수정했습니다.
- TODO  각 루트의 타입이 다르기에 Any로 설정하였습니다. 나중에 하나의 루트로 두면 좋을 것 같습니다

stopstone committed:
Merge remote-tracking branch 'origin/feat/onboarding-pages' into feat/onboarding-pages

# Conflicts:
#	Near/app/src/main/java/com/alarmy/near/presentation/feature/main/NearNavHost.kt

stopstone committed:
refactor: 온보딩 관련 중복 문자열 제거

- values/strings.xml에서 중복된 온보딩 관련 문자열 리소스를 삭제했습니다.

stopstone committed:
Merge branch 'dev' into feat/my-profile

stopstone committed:
refactor: 인증 헤더 삭제

- `MemberApiService`의 `getMyInfo`, `withdraw` 함수에서 Authorization 헤더 파라미터 삭제
- `MemberRepositoryImpl`에서 `getMyInfo`, `withdraw` 호출 시 Authorization 헤더 전달 로직 삭제

stopstone committed:
refactor: MyProfileUiState에서 사용하지 않는 error 필드 제거

stopstone committed:
refactor: RouteWebView에서 toRoute 사용하도록 수정

- 기존에 arguments에서 직접 값을 가져오던 방식에서 `toRoute<RouteWebView>()`를 사용하도록 변경했습니다.

stopstone committed:
feat: WebView 뒤로가기 기능 추가

- WebViewFrame 컴포저블에 BackHandler를 추가하여 WebView 내에서 뒤로가기 기능을 구현했습니다.
- WebView의 canGoBack 상태를 추적하여 뒤로가기 가능 여부를 결정합니다.
- 페이지 로드 완료 시 및 WebView 업데이트 시 canGoBack 상태를 업데이트합니다.

stopstone committed:
refactor: Glide의 contentScale 옵션 제거

- GlideImage의 contentScale 파라미터가 이미 존재하므로 중복되는 RequestOptions의 contentScale 관련 코드를 삭제했습니다.

stopstone committed:
fix: `NearOutlinedTextField` 에러 상태 반영 오류 수정

- `OutlinedTextFieldDefaults.Container`의 `isError` 파라미터에 `isError` 상태를 직접 전달하도록 수정하여, 텍스트 필드의 에러 상태가 올바르게 반영되도록 했습니다.

stopstone committed:
refactor: 회원탈퇴 화면 에러 처리 방식 변경

- WithdrawViewModel에서 errorMessage를 직접 관리하는 대신 WithdrawUiEvent.ShowError 이벤트를 발생시켜 UI에 에러를 전달하도록 수정했습니다.
- WithdrawScreen에서 ShowError 이벤트를 수신하여 스낵바를 통해 에러 메시지를 표시하도록 변경했습니다.
- WithdrawUiState에서 errorMessage 필드를 제거했습니다.
- NavigationMyProfile의 WithdrawRoute에 onShowErrorSnackBar 파라미터를 추가하여 에러 발생 시 스낵바를 표시할 수 있도록 했습니다.

stopstone committed:
chore: ci 워크플로우에 개인정보처리방침 URL 추가

stopstone committed:
Merge pull request #29 from near-Contact-Reminder/feat/onboarding-pages

[Feature] 앱 초기 튜토리얼 화면

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/data/mapper/MemberMapper.kt:
```diff
@@ -0,0 +1,26 @@
+ * Member 관련 데이터 변환 매퍼
+ * Data 계층 모델을 Presentation 계층 모델로 변환
+ */
+fun MemberInfo.toMyProfileInfo(): MyProfileInfo =
```
각 계층간 비슷한 모델이 쓰이면서 혼동을 줄 수 있을 것 같습니다!
적절한 접미사로 UI 레이어와 모델 레이어를 구분해보는 것은 어떨까요?
저는 현재 아래와 같은 방식으로 구성하고 있어요!
UI Layer -> ~~UIModel, ~~UIState
Model -> 접미사 x
Data Layer -> ~~Entity

stopstone replied:
앗 확실히 그런 문제점이 있을 수 있겠군요!
말씀 주신 방식으로 저도 적용해보겠습니다
감사드려요!!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/data/repository/MemberRepositoryImpl.kt:
```diff
@@ -0,0 +1,58 @@
+        override fun getMyInfo(): Flow<MemberInfo> =
+            flow {
+                emit(
+                    runCatching {
```
Flow 확장함수 통해 예외 catching 할 수 있을 것 같아요!
혹시 내부에 runCatching을 사용주신 이유가 있을까요?

stopstone replied:
예외처리를 하고자 작성했는데 불풀요한 코드였네요!
flow만 사용해도 되는 코드였습니다 ㅠ_ㅠ

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/network/service/MemberApiService.kt:
```diff
@@ -0,0 +1,24 @@
+    suspend fun getMyInfo(): Response<MemberInfo>
+
+    // 회원 탈퇴
+    @HTTP(method = "DELETE", path = "member/withdraw", hasBody = true)
```
오 Delete시에 request body가 포함되는군요..
HTTP 어노테이션 배워갑니다👍

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/myprofile/MyProfileScreen.kt:
```diff
@@ -0,0 +1,302 @@
+            color = NearTheme.colors.BLACK_1A1A1A,
+        )
+
+        NearSwitch { }
```
람다 내부가 비워져있는데, 아무 역할 없는 것 맞을까요?🧐

stopstone replied:
해당 코드에서는 스위치 배치만 하고 동작은 하지 않습니다..
TODO를 작성해두는 편이 좋았겠네요..!

알람에 대해 주형님께서 담당해주시면 그 부분에 맞춰서 통일하면 좋을 것 같아 두었었어요
지금 생각해보니 먼저 해두는 것도 좋았을 것 같습니다 🥲
알람 여부에 동의에 대해 구성하고 상태 반영 및 동작하도록 수정 예정입니다!

rhkrwngud445 replied:
알람 스위치였군요😅
이 부분은 알람 기능 구성시에 제가 함께 적용하도록 하겠습니다!

stopstone replied:
감사합니다!!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/myprofile/MyProfileViewModel.kt:
```diff
@@ -0,0 +1,129 @@
+         * 백 네비게이션 이벤트 발생
+         */
+        fun onNavigateBack() {
+            _uiEvent.trySend(MyProfileUiEvent.NavigateBack)
```
trySend를 써주셨군요! send 대신에 trySend 메소드를 사용주신 이유가 궁금합니다!

stopstone replied:
trySend는 suspend가 아닌 함수에서도 사용이 가능하여 사용하였습니다!
불필요하게 suspend가 적용되는 것 같아 trySend를 사용하였어요

trySend는 에러처리를 해주는 것이 좋지만,
에러 처리를 해주지 않은 이유는 화면전환과 같은 단발성 이벤트를 동작시켜 안전하다고 생각해서 trySend를 활용하였습니다!
(안전하지 않을수도 있어요)

rhkrwngud445 replied:
오 suspend 없이 동작하는군요! 에러처리시에는 send를 사용주셨는데, 해당 블록은 suspend 블록이어서 적용주신 것 맞을까요?

고려해야할 점이 있긴 합니다!
확률은 거의 없겠지만, 화면 회전과 같은 시나리오에서는 failure로 발생할 수 있을 것 같아요! (ex. LaunchedEffect에서 collect보다 viewModel 작업이 빠르게 되는 경우에 발현, launchedEffect에 delay를 주시면 재현할 수 있습니다.)

현재 문제가 생기는 코드가 아니니, 지석님이 선호하시는 방향으로 진행주시면 되겠습니다!

[관련 링크](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.channels/-channel/-factory/-r-e-n-d-e-z-v-o-u-s.html)

stopstone replied:
해당 케이스에 대해서는 인지하고 있습니다 :)
다만 발생 확률이 많지 않을 것 같아 현재 방식을 유지해도 좋을 것 같습니다!
피드백 감사합니다! 해당 부분은 계속 테스트 해보면서 문제가 발생하면 개선해보겠습니다 :)

rhkrwngud445 replied:
좋습니다!
조금 더 깊이 이야기 나누고 싶은 주제가 있어요!
이야기 나누었듯이 send, trySend에서 트레이드 오프가 있는 것 같아요!
|trySend|send|
|---|---|
|susepnd 제거 가능, 적은 확률로 sideEffect 발생|launch 블록 또는 susepend 구문 필요, 이벤트를 보내는 것을 보장|

저는 소수의 버그 확률이더라도 현재 코드 수정 비용 대비 sideEffect를 줄일 수 있어 send를 사용할 것 같아요! 만약 버그 발현이 된다면, 원인 파악도 비용이라고 생각해서요!
수정 반영을 요청드리는 것이 아닌, 지석님의 결정 기준은 어떻게 되는지 궁금합니다!
편하게 의견주시면 감사하곘습니다!

stopstone replied:
앗 감사합니다 확실히 send를 활용하는 것이 추후 에러방지나 유지보수 측면에서 이점에 훨씬 있는 것 같아요!
저는 단순한 이벤트를 변경하는데 코루틴을 생성해야할까? 라는 의문점과 함께 성능과 간결성 측면에서 trySend가 활용될만 하다고 생각해여 사용하고 있습니다!!

rhkrwngud445 replied:
답변 감사합니다! trySend는 성능적 이점도 가져갈 수 있겠네요!
결론지으면, 안정성이 중요한 부분에는 send 성능이 중요한 부분에는 trySend를 사용하는 것으로 결정할 수 있겠네요!

현재 코드는 수정하지 않고, 앞으로 위 기준을 바탕으로 적용하는 것으로 해요!
저도 필요한 부분에는 trySend로 적용해보겠습니다!

디스커션이 길었는데, 성심껏 답변주셔서 감사합니다!👍

stopstone replied:
즐거운 이야기 재미있었습니다! 앞으로 말씀주신대로 위 기준으로 햅보면 좋을 것 같네요!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/ui/extension/ImageLoader.kt:
```diff
@@ -0,0 +1,46 @@
+    modifier: Modifier = Modifier,
+) {
+    if (!uri.isNullOrEmpty()) {
+        GlideImage(
```
오 Glide를 써주셨군요! 저희 Coil 사용하기로 했던 것 같은데, Glide 사용주신 이유가 있을까요?

stopstone replied:
앗..... 논의 전에 작성한 코드였는데 Coil을 사용하기로 한 것을 잊고 있었습니다..!
해당 pr에서 수정후 올리겠습니다

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/myprofile/WithdrawViewModel.kt:
```diff
@@ -0,0 +1,175 @@
+
+            viewModelScope.launch {
+                val request =
+                    WithdrawRequest(
```
저는 Request를 만드는 것은 Repository 레이어 관심사에 가깝다고 생각합니다! 혹시 지석님 생각은 어떠실까요?

stopstone replied:
이 부분에 대해서는 확실히 기준을 못잡고 있긴 합니다
1. **data 계층**
API와의 통신에 필요한 데이터이니 Request는 data 계층
ViewModel은 데이터의 상태를 감지하여 UI에 반영하는 역할

2. **ViewModel 계층**
Request 내부 속성들 (ex. WithdrawRequest의 탈퇴 사유)  같은 것들에 대해 Repository에서 해줘야 할 것이 없다면
아예 ViewModel에서 처리하는 것이 추후 수정할 때도 나을 것 같다고 생각하여 ViewModel에 책임을 두었습니다!

하지만 지금은 data/model에 WithdrawRequest를 위치해 두었으니 앞뒤가 안맞는.. 코드가 되어버린 것 같습니다
Repository에서 처리하는 것이 일관성 있을 것 같아요!
감사합니다 :)

rhkrwngud445 replied:
아하 data/model에 request를 만들어주셨군요!
이 request를 network 레이어의 모델로 고려해보는 것은 어떨까요?

저는 아래 맥락으로 로직 위치를 정하고 있습니다!
"사용자는 탈퇴할 때 탈퇴이유만 필요하지만, 서버에 요청할 때 WithdrawRequest를 통해 요청해야 한다."

만약 model을 통해 맵핑하고자 한다면, data/model 이 아닌 model 패키지 내에 WithDrawForm 클래스를 만들어 직렬화 구성 및 Request로 활용해도 괜찮을 것 같네요!

정답이 없는 문제라 저희 내부의 규칙을 정해도 좋을 것 같습니다!

stopstone replied:
오!! 감사합니다 말씀주신대로 나우면 확실히 책임이 명확해지겠네요 개선해보겠습니다 ;)

rhkrwngud445 reviewed (commented):
고생하셨습니다!
큰 작업인데, 꼼꼼하게 작업 주셨군요 감사합니다😀
사실 디스코드에 웹뷰 링크가 안보여서 구동은 못해봤습니다😂 코드 위주로 확인하여 리뷰드립니다!
궁금한 점 위주로 코멘트 남겼습니다. 확인 부탁드려요!

slg1119 commented on Near/app/src/main/java/com/alarmy/near/data/repository/MemberRepositoryImpl.kt:
```diff
@@ -0,0 +1,58 @@
+                        if (response.isSuccessful) {
+                            response.body() ?: throw Exception("회원 정보가 null입니다.")
+                        } else {
+                            val errorMessage =
```
@stopstone // 코드 잘 보았습니다! 지석님이 작성하신 코드 중에 404, 401, etc를 switch하는 구문이 자주 보이는데, 이걸 한 곳에 묶어놓는건 어떨까요? 🤔

stopstone replied:
아 좋은 의견 감사합니다!
며칠 전 고민하던 내용이였는데 저 메시지들로 처리가 안되는 현상이 있어서 그 문제 해결할 때 리팩토링 해보겠습니다 :)

rhkrwngud445 reviewed (commented):
코멘트에 대한 추가 답변 드립니다!

stopstone committed:
refactor: Glide에서 Coil로 이미지 로딩 라이브러리 변경

- 기존 Glide를 사용하던 ImageLoader를 Coil을 사용하도록 수정했습니다.
- `AsyncImage`를 사용하여 이미지를 비동기적으로 로드하고, `ImageRequest.Builder`를 통해 crossfade 효과를 적용했습니다.
- crossfade는 논의 후붎필요하면 제거하겠습니다!
- 관련 의존성을 `libs.versions.toml` 및 `app/build.gradle.kts`에서 Glide에서 Coil로 변경하였습니다.

stopstone committed:
refactor: API 응답 타입 수정

- `MemberApiService`의 `getMyInfo` 함수의 반환 타입을 `Response<MemberInfo>`에서 `MemberInfo`로 변경했습니다.
- `withdraw` 함수의 반환 타입을 `Response<Unit>`에서 `Unit`으로 변경했습니다.

stopstone committed:
feat: flow를 간소화 한 apiCallFlow 추가
- `apiCallFlow` 함수를 추가하여 Repository에서 API 호출 시 예외 처리를 단순화했습니다.
- apiCallFlow를 활용하면 emit을 생략할 수 있습니다.

stopstone committed:
refactor: API 호출 Flow 적용

- 불필요한 runCatching을 제거하였습니다.
- `apiCallFlow` 함수를 추가하여 Repository에서 API 호출 시 예외 처리를 단순화했습니다.
- `MemberRepositoryImpl`의 `getMyInfo` 및 `withdraw` 함수에 `apiCallFlow`를 적용하여 중복 코드를 제거하고 가독성을 향상했습니다.

stopstone committed:
refactor: 회원 탈퇴 API 에러 핸들링 수정

- 회원 탈퇴 API 호출 시 `catch`를 사용하여 에러를 처리하도록 수정했습니다.

stopstone committed:
refactor: Member 관련 데이터 모델 네이밍 및 매퍼 구조 변경

- `MemberInfoEntity` 및 `WithdrawRequestEntity`를 추가하여 Data Layer의 API 요청/응답 모델을 정의했습니다.
- 기존 Model Layer의 `MemberInfo`, `WithdrawRequest`를 Data Layer와 Presentation Layer 사이의 중간 모델로 사용하도록 수정했습니다.
- `MemberMapper`를 개편하여 각 Layer 간 모델 변환 로직을 명확히 분리했습니다.
    - Data Layer Entity <-> Model Layer 변환 함수 추가
    - Model Layer <-> UI Layer 변환 함수명 변경 (`toMyProfileInfo` -> `toMyProfileInfoUIModel`)
- `MemberRepositoryImpl`에서 API 호출 시 Data Layer Entity를 사용하도록 수정했습니다.
- `MyProfileViewModel`에서 UI 모델명 변경사항을 반영했습니다. (`MyProfileInfo` -> `MyProfileInfoUIModel`)

stopstone committed:
refactor: `WithdrawRequest` 모델의 위치 변경 및 관련 로직 수정

- `WithdrawRequest` 모델을 `model/member` 패키지에서 `network/request` 패키지로 이동했습니다.
- `MemberRepository`의 `withdraw` 함수 시그니처를 변경했습니다.
    - 기존: `fun withdraw(request: WithdrawRequest): Flow<Unit>`
    - 변경: `fun withdraw(reason: WithdrawReason, customReason: String? = null): Flow<Unit>`

stopstone committed:
fix: 중복된 NavGraph 정의 제거
- 머지 충돌로 인하여 중복된 NavGraph를 제거합니다.

stopstone committed:
refactor: `MyProfileInfo`를 `MyProfileInfoUIModel`로 이름 변경

stopstone committed:
refactor: 불필요한 import 제거

stopstone committed:
Merge branch 'feat/my-profile' into dev

# Conflicts:
#	Near/app/src/main/java/com/alarmy/near/data/di/RepositoryModule.kt
#	Near/app/src/main/java/com/alarmy/near/presentation/feature/main/NearNavHost.kt

stopstone committed:
refactor: `NearNavHost` 시작 화면 `startDestination`으로 통합

- `NearNavHost`의 시작 화면을 결정하는 `isLoggedIn` 파라미터를 제거하고, `startDestination` 파라미터로 통합하여 외부에서 시작 화면을 직접 지정하도록 변경했습니다.
